### DISCUSSION TITLE
proposal: Embed `public_suffix_list.dat` directly without git submodules

### DISCUSSION THREAD

**Discussion by Xuanwo** [sagebind/isahc#407] [Ideas]
Status: **Accepted**

---

https://github.com/sagebind/isahc/blob/7163e8ab8b4bb86893ed8d95e977eeba797be601/.gitmodules#L1-L3

We make psl list a submodule three years ago. And I think it isn't worth it to introduce a git submodule a simple file.

`publicsuffix/list` is licensed under `MPL-2.0`, maybe we can embed the file directly in our repo with a clear license statement?

We can download from https://publicsuffix.org/list/public_suffix_list.dat directly and validate it via it's etag in md5

```shell
:) curl -I https://publicsuffix.org/list/public_suffix_list.dat
HTTP/1.1 200 Connection established

HTTP/2 200
content-type: text/plain; charset=UTF-8
content-length: 242130
last-modified: Thu, 11 Aug 2022 17:38:11 GMT
server: AmazonS3
date: Fri, 19 Aug 2022 06:55:58 GMT
etag: "8ee4364ffd37f8ec963931352799a7fb"
vary: Accept-Encoding
x-cache: Hit from cloudfront
via: 1.1 a06140ffee86972bad90c57fc682df36.cloudfront.net (CloudFront)
x-amz-cf-pop: SEA19-C3
x-amz-cf-id: PnPTxoN1CzVlmbQ1tmw6n4mFaL_F8M-YAQpHfPbsd0OKWNlqvZnFHQ==
age: 328
```

This change will make our contributors happy for the first time they are trying to build isahc locally.

Besides, it's the recommend way as described in `public_suffix_list.dat`:

```
// Please pull this list from, and only from https://publicsuffix.org/list/public_suffix_list.dat,
// rather than any other VCS sites. Pulling from any other URL is not guaranteed to be supported.
```

---

cc @sagebind for comments.

---

Updates: Implemented in https://github.com/sagebind/isahc/pull/408

**Comment by sagebind** [comment]
Yeah, this is probably a good idea as the submodule is a bit of a chore. I would not download the list directly at build time since requiring network to compile is probably worse. But since we have @teto-bot set up on a schedule to update the submodule anyway, we can reduce the tedium by committing the file directly to the repo as you suggest and have @teto-bot update it periodically via a PR. We just have to make sure to include the license info.

Note that we already download the public suffix at runtime with the recommended URL, but we _also_ provide a precompiled one in case the public suffix is not reachable at runtime. Incidentally it might be nice to allow users to configure whether the list should be updated automatically or not, and to tie it to a running `HttpClient` as well.

**Reply by Xuanwo** [reply]
Implemented in https://github.com/sagebind/isahc/pull/408, PTAL.
